libraries/heft-config-file/src/ConfigurationFile.ts

```diff
@@ -668,6 +557,248 @@ export class ConfigurationFile<TConfigurationFile> {
     }
   }
 
+  private _mergeConfigurationFiles(
+    parentConfiguration: TConfigurationFile | undefined,
+    configurationJson: IConfigurationJson & TConfigurationFile,
+    resolvedConfigurationFilePath: string
+  ): TConfigurationFile {
+    const ignoreProperties: Set<string> = new Set(['extends', '$schema']);
+    return this._mergeObjects(
+      parentConfiguration,
+      configurationJson,
+      resolvedConfigurationFilePath,
+      this._propertyInheritanceTypes,
+      ignoreProperties
+    );
+  }
+
+  private _mergeObjects<TField>(
+    parentObject: TField | undefined,
+    currentObject: TField | undefined,
+    resolvedConfigurationFilePath: string,
+    configuredPropertyInheritance?: IPropertiesInheritance<TField>,
+    ignoreProperties?: Set<string>
+  ): TField {
+    const resultAnnotation: IConfigurationFileFieldAnnotation<TField> = {
+      configurationFilePath: resolvedConfigurationFilePath,
+      originalValues: {} as TField
+    };
+    const result: TField = {
+      [CONFIGURATION_FILE_FIELD_ANNOTATION]: resultAnnotation
+    } as unknown as TField;
+
+    // Do a first pass to gather and strip the inheritance type annotations from the merging object
+    const currentObjectPropertyNames: Set<string> = new Set(Object.keys(currentObject || {}));
+    const filteredObjectPropertyNames: string[] = [];
+    const inheritanceTypeMap: Map<string, IPropertyInheritance<InheritanceType>> = new Map();
+    for (const propertyName of currentObjectPropertyNames) {
+      if (ignoreProperties && ignoreProperties.has(propertyName)) {
+        continue;
+      }
+      const inheritanceTypeMatches: RegExpMatchArray | null = propertyName.match(
+        CONFIGURATION_FILE_MERGE_BEHAVIOR_FIELD_REGEX
+      );
+      if (inheritanceTypeMatches && inheritanceTypeMatches.length === 2) {
```

Comment:
Length should always be `2`, right?

Do you think we should fail if something looks like a magic property, but isn't? We should also probably have a way to escape these. Maybe `$$` is an escaped `$`.

Reply:
It should always be `2` since the regex is intentionally tight, correct. I can remove that check since it's not really necessary.

Not so sure we should fail on things that look like magic properties though, at least not yet. If we're sticking to the `$<propertyName>.inheritanceType` style for now, then the only way to collide with this would be to be of that exact format. If we were to generalize the magic properties to be like what you described above (eg. `"$<propertyName": { ... }`) then it becomes more of a problem, and would likely necessitate some form of error case (and escaping from that error case... I like the `$$` to escape the `$` in that scenario). For now though, that doesn't really seem like a problem given the current constraints.

These are really good points though and I'm going to drop a comment in to point them out in the future, should we go that route.